Objects/dictobject.c

```diff
@@ -6819,38 +6999,95 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
     return 0;
 }
 
+static bool
+set_dict_inline_values(PyObject *obj, PyObject *new_dict)
+{
+    PyDictValues *values = _PyObject_InlineValues(obj);
+
+    if (values->valid) {
```

Comment:
The order of operations here seems strange. I think it means that concurrent reads can see empty or partially emptying dicts.

Can we assign the new dict, then set `valid` to 0, and then clear the old values?